src/RulesDeploy.ts

```diff
@@ -53,72 +58,96 @@ export class RulesDeploy {
    * Compile all rulesets tied to this deploy, rejecting on first
    * compilation error.
    */
-  compile(): Promise<any> {
-    const promises: any[] = [];
-    _.forEach(this.rulesFiles, (files: RulesetFile[], filename: string) => {
-      promises.push(this._compileRuleset(filename, files));
-    });
-    return Promise.all(promises);
+  async compile(): Promise<void> {
+    for (const filename of Object.keys(this.rulesFiles)) {
+      const files = this.rulesFiles[filename];
+      await this._compileRuleset(filename, files);
+    }
   }
 
   /**
    * Create rulesets for each file added to this deploy, and record
    * the name for use in the release process later.
+   *
+   * If the ruleset to create is identical to the latest existing ruleset,
+   * then we record the existing ruleset name instead of creating a duplicate.
+   *
+   * @param {string} service The service to create a ruleset.
+   *    Valid values are "cloud.firestore" and "firebase.storage".
```

Comment:
can we do wacky typescript things to enforce this, like `type Service = "cloud.firestore" | "firebase.storage"`?

Reply:
yes! done.